pkg/authorization/api/v1beta1/types.go

```diff
@@ -82,6 +82,40 @@ type PolicyBinding struct {
 	RoleBindings []NamedRoleBinding `json:"roleBindings"`
 }
 
+// ResourceAccessReviewSpec contains information about the action being requested.
+type ResourceAccessReviewSpec struct {
+	// Verb is one of: get, list, watch, create, update, delete
+	Verb string `json:"verb"`
+	// Resource is one of the existing resource types
+	Resource string `json:"resource"`
+	// Content is the actual content of the request for create and update
+	Content kruntime.RawExtension `json:"content,omitempty"`
+	// ResourceName is the name of the resource being requested for a "get" or deleted for a "delete"
+	ResourceName string `json:"resourceName,omitempty"`
+}
+
+// ResourceAccessReviewStatus describes who can perform the action
+type ResourceAccessReviewStatus struct {
+	// Users is the list of users who can perform the action
+	Users []string `json:"users"`
+	// Groups is the list of groups who can perform the action
+	Groups []string `json:"groups"`
+	// EvaluationError is optional.  It indicates why a ResourceAccessReview failed during evaluation
+	EvaluationError string `json:"evaluationError,omitempty"`
+}
+
+// ResourceAccessReview is a means to request a list of which users and groups are authorized to perform the
+// action specified by spec
+type ResourceAccessReview struct {
+	kapi.TypeMeta   `json:",inline"`
+	kapi.ObjectMeta `json:"metadata,omitempty" `
```

Comment:
ResourceAccessReview shouldn't have a self link at all in theory, but you should use ListMeta instead of ObjectMeta (listMeta is for non objects).

----- Original Message -----
> > +
> > +// ResourceAccessReviewStatus describes who can perform the action
> > +type ResourceAccessReviewStatus struct {
> > +	// Users is the list of users who can perform the action
> > +	Users []string `json:"users"`
> > +	// Groups is the list of groups who can perform the action
> > +	Groups []string `json:"groups"`
> > +	// EvaluationError is optional.  It indicates why a ResourceAccessReview
> > failed during evaluation
> > +	EvaluationError string `json:"evaluationError,omitempty"`
> > +}
> > +
> > +// ResourceAccessReview is a means to request a list of which users and
> > groups are authorized to perform the
> > +// action specified by spec
> > +type ResourceAccessReview struct {
> > +	kapi.TypeMeta   `json:",inline"`
> > +	kapi.ObjectMeta `json:"metadata,omitempty" `
> 
> @smarterclayton Removing ObjectMeta produces an error when trying to create a
> self-link:
> ```
> E0211 13:32:36.634201   13281 resthandler.go:131] unable to set self link for
> apiserver.RESTResult{Object:(*api.ResourceAccessReview)(0xc2087b6a00),
> Created:false}: unable to find object fields on
> reflect.Value{typ:(*reflect.rtype)(0x109b340),
> ptr:(unsafe.Pointer)(0xc2087b6a00), scalar:0x0, flag:0x196}: couldn't find
> Namespace field in api.TypeMeta{Kind:"", APIVersion:""}
> ```
> 
> I want a namespaced resource in order to have easily described access control
> on this API.  Is pretending to be an API object too high a cost?
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/800/files#r24520407